### PR TITLE
Adds support for upgrading versions from 4.7.5 to 5.x thus connecting…

### PR DESCRIPTION
**Pull request by dbernstein**
**Adds support for upgrading versions to the 4.7.5->5 path**
* * *

**JIRA Ticket**:Resolves: https://jira.lyrasis.org/browse/FCREPO-2984

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
Users can now migrate versions from 4.7.5 to 5.1.1 and by implication 6.
# What's new?
* Support for versions 4.7.5->5.x
* Enhanced integration test

# How should this be tested?

From the fcrepo-migration-utils: 
```
mkdir -p ~/tmp/fcrepo-5.1.1-with-versions-export
# upgrade 4.7.5 to 5.1.1
java -Dfcrepo.log=DEBUG -jar target/fcrepo-upgrade-utils-6.0.0-SNAPSHOT.jar -i src/test/resources/4.7.5-export  -s 4.7.5 -t 5+ -o ~/tmp/fcrepo-5.1.1-with-versions-export -u http://localhost:8080/rest
# upgrade 5.1.1 to 6.
mkdir  -p ~/tmp/fcrepo-6.0.0-ocfl-with-versions/data
java -Dfcrepo.log=DEBUG -jar target/fcrepo-upgrade-utils-6.0.0-SNAPSHOT.jar -i /Users/danielbernstein/tmp/fcrepo-5.1.1-with-versions-export-3  -o ~/tmp/fcrepo-6.0.0-ocfl-with-versions/data  -u  http://localhost:8080/rest -s 5+ -t 6+

# run the latest  fcrepo main:
mvn -Dfcrepo.home=~/tmp/fcrepo-6.0.0-ocfl-with-versions jetty:run -pl fcrepo-webapp
```
Browse the application and verify that  you can see all the resources.
Note for binaries and binary metadata you'll need to access Fedora via the command line:
```
curl -I http://localhost:8080/rest/container1/tesbinary/fcr:metadata/fcr:versions
curl -I http://localhost:8080/rest/container1/tesbinary/fcr:versions

```

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo4/committers

### CODE REVIEW

**Comment by awoods** [comment]
I am unable to test this since F4 versions appear to no longer be exported. See: https://github.com/fcrepo-exts/fcrepo-import-export/issues/155

My test process is:
1. Start F4
1. Create F4 objects/binaries with versions
1. Run import/export of F4
1. Run upgrade/utils on export -> F5
1. Run upgrade/utils on export -> F6
1. Start F6 over export

The third step above does not produce version exports... which is one of the key points of testing for this PR. 
@dbernstein : did you employ a different process for creating exported F4 versions?

**Comment by dbernstein** [comment]
@awoods : did you use fcrepo-import-export-0.3.0 to create your fcrepo 4.7.5 export?

**Comment by awoods** [comment]
Hi @dbernstein ,
No, I did not. Thanks for the link to the migration wiki. I have now moved it to be a child of: https://wiki.lyrasis.org/display/FF/Fedora+Upgrade+Notes

**Review by awoods — CHANGES_REQUESTED** [review]
Before doing a code review, I tested the upgrade functionality. There seem to be a few issues. Here are my steps:
1. Start F4
   ```
   java -jar fcrepo-webapp-4.7.5-jetty-console.jar  --headless
   ```
1. Create F4 objects/binaries with versions ([f4-test-objects.sh](https://gist.github.com/awoods/59eccb055821582176dfb10c5d0be027))
   ```
   ./f4-test-objects.sh
   ```
1. Run export of F4
   ```
   java -jar fcrepo-import-export-0.3.0.jar --mode export --resource http://localhost:8080/rest --dir ../f4-export --binaries --versions
   ```
1. Run upgrade/utils on export -> F5
   ```
   java -jar fcrepo-upgrade-utils-6.0.0-SNAPSHOT.jar -i ../f4-export/ ../f5-upgrade -s 4.7.5 -t 5+ -u http://localhost:8080/rest
   ```

Observations:
1. The only output was:
   ```
   INFO 13:22:54.980 (F47ToF5UpgradeManager) Processing directory: /home/awoods/Documents/work/fedora/notes/tasks/fedora/2020/test/scripts/upgrade-f4-to-f6/bin/../f4-export
   ```
   ..then nothing. It would be helpful to have some summary text on successful run.
1. Usage misspelling: `fcepo-upgrade-util`
1. Upgrade to 5+ creates output dir, upgrade to 6+ does not
1. Several errors from run:
   ```
   ERROR 13:35:34.267 (MigrateResourceTask) Failed to process ResourceInfo{parentId='info:fedora', fullId='info:fedora/a', nameEncoded='a', outerDirectory=../f5-upgrade/rest, innerDirectory=../f5-upgrade/rest/a, type=CONTAINER} java.time.format.DateTimeParseException: Text 'v1' could not be parsed at index 0
   ```
   ```
   ERROR 13:35:34.373 (MigrateResourceTask) Failed to process ResourceInfo{parentId='info:fedora/objects/raven/pages', fullId='info:fedora/objects/raven/pages/page1', nameEncoded='page1', outerDirectory=../f5-upgrade/rest/objects/raven/pages, innerDirectory=../f5-upgrade/rest/objects/raven/pages/page1, type=CONTAINER} java.io.UncheckedIOException: java.nio.file.NoSuchFileException: ../f5-upgrade/rest/objects/raven/pages/page1
   ```

**Comment by pwinckles** [comment]
@awoods I was just reading through your comment and one thing struck me. You ran:

`java -jar fcrepo-upgrade-utils-6.0.0-SNAPSHOT.jar -i ../f4-export/ ../f5-upgrade -s 4.7.5 -t 5+ -u http://localhost:8080/rest`

And saw exceptions like:

`ERROR 13:35:34.267 (MigrateResourceTask) Failed to process ResourceInfo{parentId='info:fedora', fullId='info:fedora/a', nameEncoded='a', outerDirectory=../f5-upgrade/rest, innerDirectory=../f5-upgrade/rest/a, type=CONTAINER} java.time.format.DateTimeParseException: Text 'v1' could not be parsed at index 0`

???

I thought `MigrateResourceTask` was just used when migrating to F6, but your command is for migrating to F5.

**Comment by awoods** [comment]
It is a good question, @pwinckles . I have not looked into where MigrateResourceTask is getting pulled in here... or why in the F5 context.

**Comment by dbernstein** [comment]
@awoods :  I'll make sure that the output directory is being created automatically in 5->6

**Comment by pwinckles** [comment]
I ran a tweaked version of @awoods's script against F5, did an export of F5, and then migrated that export to F6 without any problems.

I then did a cursory comparison of the exported F5 content against the upgraded F4-F5 content, and it appears that versions are not being handled correctly. For example:

```
# Upgraded content
$ ls f5-upgrade/rest/objects/raven/pages
cover  cover.ttl  cover.ttl.headers  page0.ttl  page0.ttl.headers  page1.ttl  page1.ttl.headers

# F5 native content
$ ls f5-export/rest/objects/raven/pages
cover  cover.ttl  page0  page0.ttl  page1  page1.ttl

$ ls f5-export/rest/objects/raven/pages/page0
fcr%3Aversions.ttl
```

Note the missing `page0/fcr%3Aversions.ttl`.

File: src/main/java/org/fcrepo/upgrade/utils/f6/ResourceMigrator.java [hunk 1]
```diff
@@ -336,6 +338,10 @@ private void deleteObject(final String fullId) {
     private Map<String, ResourceInfo> listDirectChildren(final String rootParentId,
                                                          final String currentParentId,
                                                          final Path containerDir) {
+        if(!containerDir.toFile().exists()){
```
Comment (pwinckles) [inline src/main/java/org/fcrepo/upgrade/utils/f6/ResourceMigrator.java:341]: You're missing a space after the `if` and another before the `{`

File: src/main/java/org/fcrepo/upgrade/utils/f6/ResourceMigrator.java [hunk 2]
```diff
@@ -371,8 +377,11 @@ private void deleteObject(final String fullId) {
      * @return list of ghost nodes
      */
     private List<Path> listGhostNodes(final Path containerDir, final Set<String> children) {
-        final var ghosts = new ArrayList<Path>();
+        if(!containerDir.toFile().exists()){
```
Comment (pwinckles) [inline src/main/java/org/fcrepo/upgrade/utils/f6/ResourceMigrator.java:380]: Same here

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java [hunk 3]
```diff
@@ -87,125 +113,274 @@ private void processDirectory(final File dir) {
...
+    }
+
+    private Resource getOriginalResource(Resource resource) {
+        return ResourceFactory.createResource(resource.getURI()
+                                                      .replaceAll("/fcr:versions/[a-zA-Z0-9.]*", ""));
+    }
+
+    private void addTypeLinkHeader(Map<String, List<String>> headers, String typeUri) {
+        final FcrepoLink link = FcrepoLink.fromUri(typeUri).rel(TYPE_RELATION).build();
+        headers.get("Link").add(link.toString());
```
Comment (awoods) [inline src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:308]: May as well use the `LINK_HEADER` constant you created in HttpConstants.java.

**Review by awoods — CHANGES_REQUESTED** [review]
..now to test ACLs and External Content.

**Comment by awoods** [comment]
In trying to upgrade a repository from 5->6 that was deployed to Tomcat (i.e. `/fcrepo/rest/`), I get the following error:
```
java.io.UncheckedIOException: java.nio.file.NoSuchFileException: ../f5-upgrade/rest.ttl
	at org.fcrepo.upgrade.utils.f6.RdfUtil.parseRdf(RdfUtil.java:78)
	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.readRdf(ResourceMigrator.java:500)
	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrateContainer(ResourceMigrator.java:162)
	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrate(ResourceMigrator.java:126)
	at org.fcrepo.upgrade.utils.f6.MigrateResourceTask.run(MigrateResourceTask.java:59)
```
..due to the assumption: https://github.com/fcrepo-exts/fcrepo-upgrade-utils/blob/4.7-5.x/src/main/java/org/fcrepo/upgrade/utils/F5ToF6UpgradeManager.java#L55

This issue should be worked in a separate JIRA/pull-request.

**Comment by awoods** [comment]
External content appears to work. It would be ideal if someone with an F4 with ACLs could test that ACLs are correctly migrated.

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java [hunk 4]
```diff
@@ -87,125 +113,274 @@ private void processDirectory(final File dir) {
     }
 
     private void processFile(final Path path) {
+
+        //skip versions container
+        if (path.endsWith(FCR_VERSIONS_PATH_SEGMENT + TURTLE_EXTENSION)) {
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:118]: Yes - this is true.  I would suggest we resolve in a later PR.

**Comment by dbernstein** [comment]
@awoods :  ACLs are not covered in this PR. They will not work.

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:118]: I will create a JIRA

[the remark this answers is not in the document]
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:118]: https://jira.lyrasis.org/browse/FCREPO-3521

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java [hunk 5]
```diff
@@ -87,125 +113,274 @@ private void processDirectory(final File dir) {
...
+        }
+    }
 
-                final AtomicBoolean isBinary = new AtomicBoolean(false);
-                final AtomicBoolean isExternal = new AtomicBoolean(false);
-                final AtomicBoolean isContainer = new AtomicBoolean(false);
-                final AtomicBoolean isConcreteContainerDefined = new AtomicBoolean(false);
-                final AtomicReference<Resource> containerSubject = new AtomicReference<>();
-                final AtomicBoolean rewriteModel = new AtomicBoolean();
+    private void upgradeRdfAndCreateHeaders(boolean isBinaryDescription, TemporalAccessor versionTimestamp,
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:154]: will fix.

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java [hunk 6]
```diff
@@ -87,125 +113,274 @@ private void processDirectory(final File dir) {
...
+            addTypeLinkHeader(binaryHeaders, MEMENTO.getURI());
+        }
 
-                if (isBinary.get()) {
-                    headersPrefix = newLocation.getParent().toAbsolutePath().toString();
-                    headersPrefix += isExternal.get() ? ".external" : ".binary";
-                } else {
-                    headersPrefix = newLocation.toAbsolutePath().toString();
-                }
+        // While F5 assumes BasicContainer when no concrete container is not present in the RDF on import,
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:255]: > In trying to upgrade a repository from 5->6 that was deployed to Tomcat (i.e. `/fcrepo/rest/`), I get the following error:
> 
> ```
> java.io.UncheckedIOException: java.nio.file.NoSuchFileException: ../f5-upgrade/rest.ttl
> 	at org.fcrepo.upgrade.utils.f6.RdfUtil.parseRdf(RdfUtil.java:78)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.readRdf(ResourceMigrator.java:500)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrateContainer(ResourceMigrator.java:162)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrate(ResourceMigrator.java:126)
> 	at org.fcrepo.upgrade.utils.f6.MigrateResourceTask.run(MigrateResourceTask.java:59)
> ```
> 
> ..due to the assumption: https://github.com/fcrepo-exts/fcrepo-upgrade-utils/blob/4.7-5.x/src/main/java/org/fcrepo/upgrade/utils/F5ToF6UpgradeManager.java#L55
> 
> This issue should be worked in a separate JIRA/pull-request.

https://jira.lyrasis.org/browse/FCREPO-3522

**Comment by dbernstein** [comment]
> In trying to upgrade a repository from 5->6 that was deployed to Tomcat (i.e. `/fcrepo/rest/`), I get the following error:
> 
> ```
> java.io.UncheckedIOException: java.nio.file.NoSuchFileException: ../f5-upgrade/rest.ttl
> 	at org.fcrepo.upgrade.utils.f6.RdfUtil.parseRdf(RdfUtil.java:78)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.readRdf(ResourceMigrator.java:500)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrateContainer(ResourceMigrator.java:162)
> 	at org.fcrepo.upgrade.utils.f6.ResourceMigrator.migrate(ResourceMigrator.java:126)
> 	at org.fcrepo.upgrade.utils.f6.MigrateResourceTask.run(MigrateResourceTask.java:59)
> ```
> 
> ..due to the assumption: https://github.com/fcrepo-exts/fcrepo-upgrade-utils/blob/4.7-5.x/src/main/java/org/fcrepo/upgrade/utils/F5ToF6UpgradeManager.java#L55
> 
> This issue should be worked in a separate JIRA/pull-request.

https://jira.lyrasis.org/browse/FCREPO-3522

**Comment by dbernstein** [comment]
> External content appears to work. It would be ideal if someone with an F4 with ACLs could test that ACLs are correctly migrated.

https://jira.lyrasis.org/browse/FCREPO-2985

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java [hunk 7]
```diff
@@ -298,14 +296,26 @@ private void upgradeRdfAndCreateHeaders(boolean isBinaryDescription, TemporalAcc
...
     }
 
+    private String locateBinaryHeadersPrefixForVersionedBinary(final Path newLocation) {
+        //the idea here is translate a versioned metadata resource
+        //   from
+        //   path/to/binary/fcr%3Ametdata/fcr%3Aversions/20201105171804.ttl
+        //   to
+        //   path/to/binary/fcr%3Aversions/20201105171804
+        final var fileName = newLocation.getFileName().toString();
+        final var filenameWithoutExtension = fileName.substring(0,fileName.length()-TURTLE_EXTENSION.length());
```
Comment (awoods) [inline src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java:306]: Probably less fragile to use [FilenameUtils](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FilenameUtils.html).

**Review by awoods — CHANGES_REQUESTED** [review]

File: src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java (hunk 7 above)
Comment (dbernstein) [reply src/main/java/org/fcrepo/upgrade/utils/F47ToF5UpgradeManager.java]: Done.

**Review by awoods — APPROVED** [review]

**Pull request merged**